### PR TITLE
Update/project pages

### PR DESCRIPTION
**Pull request by leannehaggerty**
# Project Pages PR Summary (`update/project_pages`)

A "shopping list" of everything introduced by this PR. For day-to-day usage
see `README.md`; for the field-by-field metadata sourcing see
`docs/metadata_source_map.md`.

## A. Overview

This PR refactors project-page `species.yaml` generation away from the legacy,
core-DB-scraping scripts (`write_yaml.py`, `hprc_write_yaml.py`) into a single
metadata-driven, auditable pipeline. Genome metadata is pulled from the Ensembl
metadata DB and the Genebuild (GB) pre-release registry, FTP publishability is
validated, icons and alternate-haplotype links are resolved, and every decision
is recorded in an audit TSV. The legacy scripts remain only for backwards
compatibility and emit a `DeprecationWarning`.

## B. Main entrypoint — `generate_project_yaml.py`

```bash
python -m ensembl.genes.projects.generate_project_yaml <input_guuids.txt> \
  --project <project> --output species.yaml --audit-file audit.tsv
```

- **Input semantics**: one identifier per line.
  - A 36-char UUID → looked up in the metadata DB (released genomes).
  - Anything else (e.g. a core DB name) → looked up in the GB registry
    (pre-release). Core-DB fallback is deprecated.
- **Discovery**: in addition to the input file, project-scoped pre-release
  genomes are auto-discovered from the GB registry (see D).
- **Output**: `--output` species.yaml (sorted by species/assembly).
- **Optional**: `--audit-file` (decision log TSV), `--changelog` /
  `--changelog-output` (diff vs a previous YAML).

## C. Metadata DB integration — `registry/metadata_db.py`

`MetadataDbClient` queries `ensembl_genome_metadata`. Fields sourced:
- species (scientific name), assembly accession, assembly name
- genome UUID (drives the beta link)
- annotation method / annotation source
- geneset last-update date
- BUSCO score + BUSCO lineage dataset
- taxonomy id and (where available) taxonomy lineage
- `get_genome_uuids_by_accessions()` — batch accession → UUID lookup used to
  turn alternate-haplotype accessions into beta links.

## D. GB registry pre-release integration — `registry/gb_tracker.py`

`GbTrackerClient` queries `gb_assembly_metadata`:
- `fetch_by_identifier()` — single core-DB/accession lookup (now also returns
  BUSCO score/lineage so icon resolution has a fallback signal).
- `fetch_project_pre_releases()` — discovers `gb_status = 'pre_released'`
  genomes scoped by `ProjectConfig` (`bioproject_scoping` /
  `custom_group_scoping`).
- Pre-release rows only become YAML rows when FTP files actually exist (see E);
  otherwise they are excluded or shown as `beta_link: Coming soon!`.

## E. FTP validation & dynamic date resolution — `yaml_renderer.py`, `ftp_client.py`

- Released FTP existence check on `ftp.ebi.ac.uk/pub/ensemblorganisms/...`.
- `annotation_source`-aware paths (e.g. `ensembl`, `braker`).
- Dynamic `YYYY_MM` date resolution: if the metadata date does not match the
  public FTP, the directory is listed and the newest matching date is used.
- Pre-release FTP fallback (`pub/databases/ensembl/pre-release/...`).
- A genome with no valid released or pre-release files is **excluded**
  (`audit_decision = excluded`).

## F. Beta link validation — `ftp_client.check_beta_species_status`

- Fetches the beta species page and inspects the **body**, because the beta
  "species not recognised" page still returns HTTP 200.
- Treated as unavailable on non-200 or on known markers
  ("We do not recognise the species identified",
  "Find available species in the Species selector").
- Only a confirmed page yields a real `beta_link`; otherwise
  `beta_link: Coming soon!`.
- **Never excludes** a genome that has valid FTP files.
- Pre-release genomes skip the check; results are cached per UUID.

## G. Changelog & audit — `changelog.py`, `--audit-file`

- `--audit-file`: one row per candidate with decision + reason, including
  `image_source`, `beta_link_status`, final `beta_link`, `taxon_id`,
  `busco_lineage`, `taxonomy_lineage`, and final `image`.
- `--changelog <old.yaml>`: human-readable added/removed/modified report;
  `--changelog-output` writes the machine-readable TSV.
- Decisions include: `included_released`, `included_prerelease`, `excluded`,
  `excluded_duplicate`, `kept_duplicate`.

## H. Duplicate handling — `generate_project_yaml.py`

- Candidates are grouped by accession (+ species + source for non-HPRC).
- Within a group the best record is kept based on date match and release
  status; the rest are marked `excluded_duplicate` / `kept_duplicate`.
- Guarantees no duplicate YAML rows for the same accession.

## I. Icon resolution — `icon_resolver.py`

- Resolves an icon from taxonomy lineage, strongest source first:
  1. metadata DB taxonomy lineage, 2. NCBI Entrez lineage (cached by taxon_id),
  3. BUSCO lineage hint (e.g. `eudicotyledons_odb12` → Plants.png), 4. default
  `Metazoa.png`.
- Most-specific match wins (Lepidoptera > Arthropoda; Mammalia/Aves/Fish/
  Reptiles/Amphibia > Chordata).
- `icons.txt` provides project overrides on top of built-in taxonomy anchors.
- Icons are **never** inferred from species names or free-text fields.

## J. Alternate haplotypes — `haplotype_resolver.py`

- `find_alternate_haplotypes()` pairs assemblies within the dataset using:
  1. shared NCBI BioSample accession (primary),
  2. shared sample/isolate name within the same taxon (secondary),
  3. assembly-name patterns (hap1/hap2, mat/pat, primary/alternate) — weak
     fallback, same-species only.
- The alternate accession is resolved to its genome UUID (local cache, then a
  batch metadata-DB lookup) and emitted as a beta link
  (`alternate: https://beta.ensembl.org/species/<uuid>`).
- If no UUID exists the `alternate` field is **omitted** (no raw accession, no
  broken link). Only assemblies present in the dataset are linked.

## K. Project-specific output support — `config.py`, `yaml_renderer.py`

- **Standard** projects (CBP, BGE, DToL, VGP, ASG, ERGA, AEGIS): image,
  accession, annotation method, FTP links, BUSCO, beta link, alternate.
- **HPRC** (`--project hprc`): `assembly_submitter`, `assembly_link`,
  `parent_of_origin` (from the NCBI assembly report), and `variants_vep`
  when present on FTP.
- **mouse_genomes** (`--project mouse_genomes` / `mouse`): explicit `strain`
  field and `alternate` haplotype linkage.

## L. Tracking script changes — `tracking/bioproject_tracking.py`

- Candidate-discovery tool (NOT the YAML-eligibility authority).
- `--project_name hprc` discovers HPRC release-2 GCAs from the HPRC catalog.
- Optional `--ftp` adds `ftp_status` columns; `--pre_release` adds GB matches.
- Config is now loaded lazily and resolved relative to the module, so the
  module imports and `--help` work from any working directory.

## M. Configuration — `config.py`, `server_config.json`

- `ProjectConfig` directs data sources, schema type, and pre-release scoping.
- `get_project_config()` maps project keys → config (hprc, mouse_genomes/mouse,
  aquafaang, vgp, darwin_tree_of_life, erga, cbp, bge, asg, and a sensible
  default for other standard projects such as AEGIS).
- `server_config.json` holds read-only DB connection details (`meta_beta`,
  `gb1`, etc.).

## N. Tests — `tests/ensembl/genes/projects/`

| File | Covers |
| --- | --- |
| `test_changelog.py` | YAML diff: added/removed/modified, date/slash normalisation, HPRC key handling |
| `test_icon_resolver.py` | metadata/NCBI/BUSCO source priority, most-specific match, plant BUSCO tokens, lineage normalisation, pre-release plant regression |
| `test_haplotype_resolver.py` | BioSample / sample-name / naming pairing, priority, species-safety, edge cases |
| `test_beta_link.py` | beta status (available/unavailable/error), caching, `_resolve_beta_link` branches, end-to-end render |

All tests mock network/DB, so they run offline. **Not covered locally** (require
cluster/internal access): live metadata DB / GB registry queries, live EBI FTP
existence checks, live NCBI Datasets/Entrez lookups, and live beta.ensembl.org
validation. Run the commands in section O on the cluster to exercise those.

## O. Validation commands (cluster)

```bash
export PYTHONPATH=$PWD/src/python

python -m ensembl.genes.projects.generate_project_yaml ./cbp_guuids.txt \
  --project cbp   --output cbp_species.yaml   --audit-file cbp_audit.tsv

python -m ensembl.genes.projects.generate_project_yaml ./bge_guuids.txt \
  --project bge   --output bge_species.yaml   --audit-file bge_audit.tsv

python -m ensembl.genes.projects.generate_project_yaml ./hprc_guuids.txt \
  --project hprc  --output hprc_species.yaml  --audit-file hprc_audit.tsv

python -m ensembl.genes.projects.generate_project_yaml ./mouse_guuids.txt \
  --project mouse_genomes --output mouse_species.yaml --audit-file mouse_audit.tsv
```

Spot-check examples:

```bash
grep -A8 "Geum rivale" bge_species.yaml          # plant icon
grep "GCA_964205265.1" bge_audit.tsv             # image_source / busco columns
grep -A15 "Achillea ptarmica" aegis_species.yaml # beta_link "Coming soon!" case
```

## P. Known limitations

- Live metadata DB, GB registry, EBI FTP, NCBI and beta.ensembl.org checks all
  require network / internal access; they cannot be exercised in offline CI.
- NCBI / Datasets lookups can fail or rate-limit; the pipeline falls back
  (e.g. BUSCO-based icon resolution) where possible and logs the path taken.
- Output YAML still depends on the `projects.ensembl.org` species.yaml table
  conventions (field names such as `image`, `beta_link`, `alternate`).

### CODE REVIEW

File: src/python/ensembl/genes/projects/.gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+*_test_yaml
```
Comment (ens-ftricomi) [inline src/python/ensembl/genes/projects/.gitignore]: can you move them into the gitignore in the root repo?

File: pyproject.toml [hunk 2]
```diff
@@ -58,7 +60,6 @@ core_meta_data = "ensembl.genes.metadata.core_meta_data:main"
 busco_lineage_selector = "ensembl.genes.metrics.busco_lineage_selector:main"
 busco_metakeys_patch = "ensembl.genes.metrics.busco_metakeys_patch:main"
 check_busco_score = "ensembl.genes.metrics.check_busco_score:main"
-write_yaml = "ensembl.genes.projects.write_yaml:main"
```
Comment (ens-ftricomi) [inline pyproject.toml:61]: can you add here the main script to call the project pages generation within the package?

File: src/python/ensembl/genes/tracking/README.md [hunk 3]
```diff
@@ -2,82 +2,128 @@
...
-3. A valid `bioproject_tracking_config.json` configuration file, which should include:
-   - MySQL connection details for your Ensembl metadata database (host, user, port, db_name).
-   - The base URLs for the NCBI Datasets API queries (e.g., for BioProject and Taxon).
-4. Install the Python dependencies listed in `requirements.txt`:
+1. **Python 3.11+** (matching the repository-wide requirement in `pyproject.toml`).
+2. **MySQL** connectivity to the Ensembl metadata database (read-only access via `ensro`).
+3. A valid `bioproject_tracking_config.json` in the working directory, containing:
+   - MySQL connection details for metadata and pre-release servers.
+   - Base URLs for the NCBI Datasets API.
+4. **NCBI Datasets CLI** (`datasets`) is optional but recommended — it retrieves all assembly versions, whereas the API fallback only returns the latest.
```
Comment (ens-ftricomi) [inline src/python/ensembl/genes/tracking/README.md:23]: there is py package https://pypi.org/project/ncbi-datasets-pyclient/, did you try it?

File: src/python/ensembl/genes/tracking/README.md [hunk 4]
```diff
@@ -2,82 +2,128 @@
...
+- `server_details`: MySQL connection details for `meta` (beta metadata DB), `pre-release` (GB1 server), and others.
+- `urls.datasets`: Base URLs for NCBI Datasets API endpoints (bioproject and taxon).
+
+**Important**: Run the script from a directory containing `bioproject_tracking_config.json`, or update the config loading path. Database passwords are typically omitted when running from the EBI cluster with standard `.my.cnf` credentials.
+
+## Usage
 
 ```bash
-python bioproject_tracking.py --haploid PRJEB40665 --classification --rank class --report_file ./dtol_rank_report_file.csv
+python -m ensembl.genes.tracking.bioproject_tracking -h
```
Comment (ens-ftricomi) [inline src/python/ensembl/genes/tracking/README.md:52]: if it is installed as a package thanks to the toml you can call directly bioproject_tracking

**Review by ens-ftricomi — COMMENTED** [review]
It looks good to me, I would decide if this section can go into the package, in this way when we use this code in other pipelines it can be mounted in an image. 
It can be done even later when we will assess the use case in Prefect

File: src/python/ensembl/genes/projects/generate_project_yaml.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+                    logger.info(
+                        "No genome UUID found for alternate %s of %s; "
+                        "omitting alternate field.",
+                        alt_acc,
+                        acc,
+                    )
+
+    if args.audit_file:
+        seen_audit_rows = set()
+        with open(args.audit_file, "a", encoding="utf-8") as af:
```
Comment (JackCurragh) [inline src/python/ensembl/genes/projects/generate_project_yaml.py]: To avoid issues f same auditfile pth is given to two different runs we should make this 'w' I believe

File: src/python/ensembl/genes/projects/config.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+def get_project_config(project_name: str) -> ProjectConfig:
+    """Factory to retrieve configuration for a specific project."""
+    name_lower = project_name.lower()
+
+    if name_lower == "hprc":
+        return ProjectConfig(
+            project_name="hprc",
+            allow_core_db_fallback=False,
+            allow_beta_urls=False,
```
Comment (JackCurragh) [inline src/python/ensembl/genes/projects/config.py]: Why not allow beta urls?

File: src/python/ensembl/genes/projects/yaml_renderer.py [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+        if meta.busco_lineage:
+            doc["busco_lineage"] = meta.busco_lineage
+
+        # Relationships
+        if meta.alternate_of:
+            doc["alternate"] = meta.alternate_of
+
+        return {k: v for k, v in doc.items() if v is not None}
+
+    def _render_hprc(self, meta: GenomeMetadata) -> Dict[str, Any]:
```
Comment (JackCurragh) [inline src/python/ensembl/genes/projects/yaml_renderer.py:398]: Even though the beta urls are turned off in the project config they will be rendered here regardless. I think

File: src/python/ensembl/genes/projects/registry/metadata_db.py
Comment (JackCurragh) [inline src/python/ensembl/genes/projects/registry/metadata_db.py:1]: May not be for this PR but we should consider using the metadata API for this so we dont have redundancy 
https://github.com/ensembl/ensembl-metadata-api/

Some things may not be in the metadata API yet and could be moved there directly and other things like getting genomes from accession are already supported. 

So I reckon this stuff should either be using the metadata API directly or wrapping around its primatives.

File: src/python/ensembl/genes/projects/ftp_client.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            )
+            file_name_protein = prod_name + "_protein_busco_short_summary.txt"
+            file_name_alternative = prod_name + "_busco_short_summary.txt"
+        else:
+            return ""
+
+        try:
+            ftp_connection = self._require_ftp(ftp_connection)
+            self.return_to_root(ftp_connection)
+            self._retry(ftp_connection.cwd, which, path)
```
Comment (JackCurragh) [inline src/python/ensembl/genes/projects/ftp_client.py:164]: I dont know if this will ever recover conenction issues because it seems to be retrying with the same potentially failed connection?

**Review by JackCurragh — COMMENTED** [review]
For the most part, this is good to go. We shoould consider standardising how we interact with the metadata DB and the registry but that is not for this PR. 

Francescas point about adding it as a subcomand in to pyproject.toml is the main one I can see that should definitely be fixed before merging. Things like the existing ncbi-datasets should also be coordinated across the whole repo rahter than just this PR.

File: src/python/ensembl/genes/tracking/README.md (hunk 3 above)
Comment (leannehaggerty) [reply src/python/ensembl/genes/tracking/README.md:23]: Nope that's a job for the development team

File: src/python/ensembl/genes/projects/ftp_client.py (hunk 8 above)
Comment (leannehaggerty) [reply src/python/ensembl/genes/projects/ftp_client.py:164]: Thanks Jack, looks like this is a real bug. No catastrophic, so I'm going to defer this to my next round of updates.

**Review by JackCurragh — APPROVED** [review]
Looks like everyting was addressed or intentionally deferred
